ldes-server-compaction/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/compaction/application/services/PaginationStartingNodeIteratorImpl.java

```diff
@@ -0,0 +1,52 @@
+package be.vlaanderen.informatievlaanderen.ldes.server.compaction.application.services;
+
+import be.vlaanderen.informatievlaanderen.ldes.server.compaction.domain.services.CompactableFragmentPredicate;
+import be.vlaanderen.informatievlaanderen.ldes.server.fragmentation.entities.Fragment;
+import be.vlaanderen.informatievlaanderen.ldes.server.fragmentation.repository.FragmentRepository;
+import be.vlaanderen.informatievlaanderen.ldes.server.domain.model.TreeRelation;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class PaginationStartingNodeIteratorImpl implements PaginationStartingNodeIterator {
+	private final FragmentRepository fragmentRepository;
+	private final CompactableFragmentPredicate compactableFragmentPredicate;
+	private final List<Fragment> structuralFragments;
+	private final List<Fragment> startingNodeFragments;
+
+	public PaginationStartingNodeIteratorImpl(FragmentRepository fragmentRepository, Fragment rootFragment) {
+		this.fragmentRepository = fragmentRepository;
+		this.compactableFragmentPredicate = new CompactableFragmentPredicate();
+		this.structuralFragments = new ArrayList<>();
+		this.structuralFragments.add(rootFragment);
+		this.startingNodeFragments = new ArrayList<>();
+	}
+
+	@Override
+	public Fragment next() {
+		return startingNodeFragments.remove(0);
+	}
+
+	@Override
+	public boolean hasNext() {
+		while (startingNodeFragments.isEmpty() && !structuralFragments.isEmpty()) {
```

Comment:
Strictly subjective but personally I prefer "apache.commons.lang3.ObjectUtils.isNotEmpty" as it reads easier than !...isEmpty

Reply:
That is indeed subjective.